### PR TITLE
Add contextMode option to wrap or remove contextTypes and childContextTypes

### LINKED ISSUE #77: PropTypes in contextTypes and childContextTypes
Could this transform also do something with `contextTypes` and `childContextTypes`?

They're the only other places `PropTypes` are used AFAIK, so handling them too would allow complete removal of `PropTypes` from an app's own production built code. This would also be useful for slimming down the production build of libraries which use context while still allowing PropTypes to be validated in development.

React doesn't care which values they have in production, but uses their property names to validate that only expected child context is provided and to mask context properties which are made available to a child, so they should be wrapped by default rather than removed (although you might still want to be able remove them if doing a cross-compatible build for something like Preact, which implements context without `contextTypes` and `childContextTypes`).

Example input and expected output:

```jsx
class Foo1 extends React.Component {
  static contextTypes = {
    bar1: React.PropTypes.string,
  };

  render() {}
}
```
```jsx
class Foo1 extends React.Component {

  render() {}
}

process.env.NODE_ENV !== "production" ? Foo1.contextTypes = {
  bar1: React.PropTypes.string
} : {
  bar1: 1
};
```

### PR DESCRIPTION
**Pull request by YasserKaddour**
This PR introduce the opt-in option `contextMode` to set up how to deal with `contextTypes` and `childContextTypes`, same as the `mode` option it can take `remove` or `wrap` but do nothing by default.

I think this would help library authors using the `context` API deal with the  [#PropTypepocalypse](https://twitter.com/timdorr/status/852219173359034375) and remove the `prop-types` dependency from production build :smile:.

I think I made too many tests, but I will let you be the judge of that.

Tell me if there is anything I overlooked :v:.

Closes #77.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -130,6 +130,24 @@ import PropTypes from 'prop-types'
 ```
  - `false` (default): does not remove the import statements.
 
+### `contextMode`
```
Comment (oliviertassinari) [inline README.md]: 👍  for an option as it can allow people to try the feature out.

File: src/remove.js [hunk 2]
```diff
@@ -19,7 +19,6 @@ export default function remove(path, globalOptions, options) {
   const {
     visitedKey,
     wrapperIfTemplate,
-    mode,
```
Comment (oliviertassinari) [inline src/remove.js]: Why?

File: test/fixtures.spec.js [hunk 3]
```diff
@@ -92,7 +93,78 @@ describe('fixtures', () => {
               };
               break;
 
+            case 'remove-context-wrap-es6':
```
Comment (oliviertassinari) [inline test/fixtures.spec.js]: Why adding those two type of tests? I don't think that we need those. I would rather use on of the following options:
- Use the `options.json` file (inspiration taken from Babel
- Enable the `contextMode` option for the tests

File: test/fixtures/es-class-assign-property/expected-remove-context-wrap-es6.js [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+
+process.env.NODE_ENV !== "production" ? Foo1.childContextTypes = {
+  bar1: PropTypes.string
+} : void 0;
+
+class Foo2 extends React.PureComponent {
+  render() {}
+}
+
+process.env.NODE_ENV !== "production" ? Foo1.contextTypes = {
```
Comment (oliviertassinari) [inline test/fixtures/es-class-assign-property/expected-remove-context-wrap-es6.js]: I don't get it, this is a breaking change. How is this related to #77? Sounds like a different use case. I'm not sure we want that output.

Comment (YasserKaddour) [inline test/fixtures/es-class-assign-property/expected-remove-context-wrap-es6.js]: This is a bad copy pasta on my part :sweat_smile: - Fixed

File: README.md (hunk 1 above)
Comment (YasserKaddour) [inline README.md]: :smile:

File: src/remove.js (hunk 2 above)
Comment (YasserKaddour) [inline src/remove.js]: Mode is no longer a global option because the value change depending on `contextMode` and the `key` being a `contexTypes` or not. see:

``` js
 mode: isKeyContextTypes ? contextMode : mode,
```

File: test/fixtures.spec.js (hunk 3 above)
Comment (YasserKaddour) [inline test/fixtures.spec.js]: I am not sure on which cases, I should test `contextMode` on. So I did it on a bunch of them. 
Could you be more explicit on how you want me to proceed.

File: test/fixtures/es-class-assign-property/expected-remove-context-wrap-es6.js (hunk 4 above)
Comment (oliviertassinari) [inline test/fixtures/es-class-assign-property/expected-remove-context-wrap-es6.js]: Sorry but I wasn't referring to the typo, I was referring to removing everything.

Comment (YasserKaddour) [inline test/fixtures/es-class-assign-property/expected-remove-context-wrap-es6.js]: Sorry, but I don't get what you are referring to :confused: , I think this is the correct behavior. This remove the `PropTypes` property, but only wrap `ContextTypes` and `childContextTypes`.

Comment (oliviertassinari) [inline test/fixtures/es-class-assign-property/expected-remove-context-wrap-es6.js]: The point is that React needs the context type key names to provide them to the component. By removing them, you lose the context values, hence, you break the component. At least, that's my understanding of the reality. Have a look at the issue, I think that it's explain in the thread.

Comment (YasserKaddour) [inline test/fixtures/es-class-assign-property/expected-remove-context-wrap-es6.js]: :sweat: I didn't get that. I will give it another shot. Sorry about that.

Comment (oliviertassinari) [inline test/fixtures/es-class-assign-property/expected-remove-context-wrap-es6.js]: I have been in that situation so many times, I hate reading the doc 😆 .

**Comment by azizhk** [comment]
Hi,
I think we need contextTypes in production.
> If `contextTypes` is not defined, then `context` will be an empty object.
From: https://facebook.github.io/react/docs/context.html#how-to-use-context

If library authors are using contextTypes they need it and without it context would be an empty object.

**Comment by oliviertassinari** [comment]
> If contextTypes is not defined, then context will be an empty object.

https://facebook.github.io/react/docs/context.html#how-to-use-context

@azizhk I have been playing with the API around. We need more than `contextTypes = {}` to get  the context values.

I'm closing the PR as has been inactive for quite some time. We still have the issue as reference.
@YasserKaddour Thanks for the effort.

**Comment by azizhk** [comment]
@oliviertassinari You tried with react or preact?
> Components do not implement contextTypes or childContextTypes. Children receive all context entries drawn from getChildContext().
> From: https://preactjs.com/guide/differences-to-react

The last line.

**Comment by oliviertassinari** [comment]
I was playing with React. Thanks for the clarification :).

**Comment by ChisholmKyle** [comment]
I'm not sure if this is still relevant, but for the `contextTypes` static property, I was able to replace `PropTypes.object` with an empty function `() => {}` to trigger the use of `context` and my app still seems to work. Note this is untested. I'm using Typescript so I get static type-checking (and run-time prop-types is no longer required). Here is an example (ignore the naming ):

**Before**
```tsx
import PropTypes from 'prop-types';

interface ProviderContext {
  io: SomeObjectType;
}

class NewComponent extends Component<void, void> {
  context: ProviderContext;
  static contextTypes = {
    io: PropTypes.object.isRequired
  };
  render() {
    return (
      <ExampleComponent {...this.props} io={this.context.io} />
    );
  }
}
```

**After**
```tsx
interface ProviderContext {
  io: SomeObjectType;
}

class NewComponent extends Component<void, void> {
  context: ProviderContext;
  static contextTypes = {
    io: () => {}
  };
  render() {
    return (
      <ExampleComponent {...this.props} io={this.context.io} />
    );
  }
}
```

**Comment by oliviertassinari** [comment]
@ChisholmKyle Thanks for sharing.
